New Loading Logic / Restore Timeline Position

S3basuchian opened the pull request:
Hi,
I propose this new loading logic which has multiple advantages over the existing logic:

- Only one loading method is needed (no need for separate `loadInitial()` method)
- If the last read tweet is within the newest 200 tweets it uses one less call to the Twitter API (since the separation between read & not yet read is done locally)
- The app can restore the last read position from a refresh within the app
- The last read position can be restored even when the last read tweet is not within the newest 200 tweets

It's my first time contributing to an open source project, so I'm hoping there are no "red flags" within my changes. If there is anything that I need to adapt/change let me know :)

robertodoering reviewed (commented):
Thanks for the PR!

It's a creative solution and the reduced request is nice.
Restoring the position on refresh is also nice but it's still quite janky - in the future I would love to find a better solution where the timeline doesn't jump at all but that's not as straight forward.

I left 2 small comments that you can address if you'd like and then we can merge it.

---

There are a few more issues that I found but they can be addressed separately (unless you want to handle them):
- Refreshing will only get the timeline position restored if the 'restore home timeline position on app start' option is enabled
  - Restoring on refresh should get its own setting
- Refreshing shouldn't use the `restoredTweetId` when restoring the position since it isn't guaranteed to be the newest tweet. Instead we should just use the first tweet in the list
- When refreshing the timeline, the text 'new tweets since last visit' should probably just me 'new tweets'
- I find the `load` function to be quite messy and hard to read. I can see why you merged the `loadInitial` and `load` since they now share the same restoring logic but this can probably be cleaned up a bit, especially when considering the new settings and `restoredTweetId`.

S3basuchian commented:
Thanks for the feedback! Couple points:

- I now decrement the `maxId` in the while loop as you suggested
- The `tweets.indexOf()` value is now saved as you suggested
- I added a new setting for restoring the position after refresh
- Refreshing now uses the id of the newest tweet in the existing timeline
- I split `load()` and `loadInitial()` again and added `loadAndRestore()` which is called from `loadInitial()` of `load()` depending on the settings and with the tweet id of the tweet that acts as a "breakpoint"
- For the text change ('new tweets since last visit') I was not quite sure how to handle this, since currently in this widget we have no way of knowing if it was an initial load or a timeline refresh. I could use different values for both instead of using `initialResultsCount` etc. but this seems kind of redundant to me, as they would serve the same purpose other than the small text change. Do you have any suggestions regarding this?

I'm not sure if my changes made the code more readable for you. If you have any suggestions how I could improve it I am happy to implement them.

Again, thanks for your great feedback thus far!

S3basuchian commented:
Thank you! Happy to help - it's the most beautiful Twitter client I have found by far!
